docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,189 @@
+
+### Use Cases
+* I want to sync all groups I've configured in a sync job (e.g. a specific LDAP server configuration) to OpenShift `Groups`
+* I want to sync specific OpenShift `Groups` with their `externalRecordProviders`
+
+### Data Flow
+* Determine list of LDAP groups to be synced
+* Find LDAP group members
+* Determine OpenShift `Identities` from LDAP group members
+* Determine OpenShift `Users` from OpenShift `Identities`
+* Determine OpenShift `Group` from LDAP group
+* Populate new OpenShift `Group` from resulting OpenShift `UserIdentityMappings` and old OpenShift `Group` fields
+
+The sync command will be exposed as `oadm sync <group(s)> [<group-name>] [filters]` and invoked like `oadm sync group <name>` or `oadm sync groups [--label=selector]`.
+
+## Support for LDAP Group Schemas
+The two main LDAP schemas used to denote groups are 1) as first-class LDAP entries and 2) as LDAP attributes on LDAP first-class entries. In the first schema, members of the LDAP group would be listed in the first-class entry as attributes. In the second schema, LDAP first-class entries for users would list the groups that they are a member of in an attribute. These groups may or may not exist as first-class entries themselves.
+
+## Sync Process
+The `LDAPMemberLister` inside of the `LDAPGroupSyncer` will, for each group that needs to be synced, find the list of entries in the LDAP server that encode the members of the group. The `LDAPCachedMemberLister` has the same functionality for LDAP sync jobs where it is not possible to run an LDAP query to get the members of each group to be synced: when syncing all groups for an LDAP schema where groups are attributes, *etc*. The `LDAPMemberIdentityFactory` creates OpenShift `Identities` for these members. The `UserIdentityMapper` determines the OpenShift `Users` that these `Identities` refer to. Syncing all or some of either LDAP or OpenShift groups is supported.
+
+```go
+type LDAPGroupSyncer struct {
+  options            LDAPSyncOptions
+  filterGenerator    LDAPFilterGenerator
+  memberLister       LDAPMemberLister
+  cachedMemberLister LDAPCachedMemberLister
+  identityFactory    LDAPMemberIdentityFactory
+  identityMapper     *userapi.UserIdentityMapper
+  client             *osclient.Client
+}
+
+func (s *LDAPGroupSyncer) SyncAllInternalGroups() (err error) {
```

Comment:
Right, so if we make the `GroupSyncer` interface we're trying to support different sources for sync - you'd commented earlier that this might not be necessary?